### ISSUE TITLE
Please add Autostart and set Port fixed

### ISSUE THREAD

**Issue by RichardBeisser** [Smooth-E/wireless-adb-switch#8]
Hello,

today I installed your WADS.
It would be great if you add 2 Features:
- Autostart while Boot
- set a fixed Port for ADB

Thanks.

**Comment by Smooth-E** [comment]
Hello. Got to say that I find your suggestions pretty reasonable. However, I should say that starting wireless debugging on boot will only be possible in the root mode, because you need to manually activate Shizuku on every boot by connecting your device to the PC. Assigning a static port may be not possible (unlikely), but I will investigate this when I get to work on the next release (in a couple of months).

**Comment by RichardBeisser** [comment]
> ... starting wireless debugging on boot will only be possible in the root mode... 

yes, i know that. all my devices are rooted with magisk.

by the way, there is already with the mentioned features, but the project seems to be maintained. 
https://github.com/RikkaApps/WADB
this App doesnt work on all of my devices, but your's does.

**Comment by Smooth-E** [comment]
> _... set a fixed Port for ADB_

I am back to working on this application for some time. Am currently a bit puzzled on whether it is even possible to set the port to some fixed value. The process of enabling wireless adb goes as follows as far as I understand by now:

1. [`AdbDebuggingManager.setAdbEnabled(...)`](https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/services/core/java/com/android/server/adb/AdbDebuggingManager.java;l=1669;drc=a78d762ccdbcb1f8f40f2c860caca75ade5d486b;bpv=0;bpt=1) is called.
2. Then the manager [sends a message](https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/services/core/java/com/android/server/adb/AdbDebuggingManager.java;l=1674;drc=a78d762ccdbcb1f8f40f2c860caca75ade5d486b;bpv=0;bpt=1) with `AdbDebuggingHandler.MSG_ADBDWIFI_DISABLE`.
3. [AdbDebuggingHandler](https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/services/core/java/com/android/server/adb/AdbDebuggingManager.java;l=654;drc=a78d762ccdbcb1f8f40f2c860caca75ade5d486b) catches the [message](https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/services/core/java/com/android/server/adb/AdbDebuggingManager.java;l=1067;drc=a78d762ccdbcb1f8f40f2c860caca75ade5d486b) to actually enable wireless debugging.
4. After checking that you are connected to a valid network, it [starts](https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/services/core/java/com/android/server/adb/AdbDebuggingManager.java;l=654;drc=a78d762ccdbcb1f8f40f2c860caca75ade5d486b) the [ConnectionPoller](https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/services/core/java/com/android/server/adb/AdbDebuggingManager.java;l=344;drc=a78d762ccdbcb1f8f40f2c860caca75ade5d486b;bpv=0;bpt=1) which monitors the system property that should contain a port for a maximum of 10 seconds and fires back the listener with what it read from the property.

So for now it looks like the port is set internally when you connect to the network. I will dig deeper a bit later, but it is possible that I will abandon this idea.

**Comment by yochananmarqos** [comment]
@Smooth-E 

> ...manually activate Shizuku on every boot by connecting your device to the PC. 

There's a Wireless debugging toggle in Developer Options in recent versions of Android. I'm not sure when it was actually added, but it's been there officially since at least Android 12 on my Pixel 6. I had a OnePlus 6T and 7 Pro just before that. I think OnePlus added it at some point while I had my 7 Pro before it was available officially.

Either way, I don't believe there's any way to enable it on boot without root. I know I've seen Tasker users request the same thing.

**Comment by Smooth-E** [comment]
Hello @yochananmarqos!

> There's a Wireless debugging toggle in Developer Options in recent versions of Android. I'm not sure when it was actually added...

I know this. It was added in Android 11 with addition of the wireless debugging feature itself.

> There's a Wireless debugging toggle in Developer Options in recent versions of Android. I'm not sure when it was actually added, but it's been there officially since at least Android 12 on my Pixel 6. I had a OnePlus 6T and 7 Pro just before that. I think OnePlus added it at some point while I had my 7 Pro before it was available officially.

What is the point of this statement exactly? Excuse me, but I don't understand what you are trying to imply here. Be reminded that WADBS (my app) is just a widget provider for a system feature. I am not implementing wireless debugging myself.

> Either way, I don't believe there's any way to enable it on boot without root. I know I've seen Tasker users request the same thing.

Yeah, and it's fine. I actually mentioned this [here](https://github.com/Smooth-E/wireless-adb-switch/issues/8#issuecomment-1861875044) as you said.

Thanks for being interested in WADBS!

**Comment by yochananmarqos** [comment]
@Smooth-E  Nevermind, I guess I misunderstood.

**Comment by madhuraj0** [comment]
Yes if possible please add ability to set a fixed port using Shizuku, currently the apps that are able to achieve it use root.

**Comment by Smooth-E** [comment]
> Yes if possible please add ability to set a fixed port using Shizuku, currently the apps that are able to achieve it use root.

Since I've taken a long break from developing WADBS for personal reasons, could you please provide example of such apps? I remember when I was researching the topic, I didn't quite get how to set a specific port even with root. Those apps will serve as a good example to start from

**Comment by madhuraj0** [comment]
This app always starts wireless debugging at 5555. However it doesn't have any option to manually set the fixed port.
https://www.apkmirror.com/apk/metactrl/wifi-adb-debug-over-air/

**Comment by Smooth-E** [comment]
> This app always starts wireless debugging at 5555. However it doesn't have any option to manually set the fixed port. https://www.apkmirror.com/apk/metactrl/wifi-adb-debug-over-air/

I think this is because apps like this use the protocol of wireless adb, but start and manage the daemon themselves. Here's a similar app by the developer of Shizuku:

https://github.com/RikkaApps/WADB

It enables wireless debugging on devices running older versions of Android, by manually implementing the wireless protocol. Since the implementation is their own, they are free to start it on whatever port they wish to.

My application, on the other hand, is just a shortcut, that triggers system properties and switches to start and manage the system-provided wireless adb implementation. This approach is much more limiting, but it is the intent of the app - to be just a handy shortcut.

As far as I remember, when the option to start wireless adb is triggered, the adb manager class refers to the system to allocate a port. Thus, the only thing I can do is hook into the system process somehow, but this will require a much stronger integration and will be much more vendor-dependent. So, ever if WADBS will someday gain such ability, it will most likely require root, because the only thing Shizuku allows is to use system protected APIs, not to hook into how they behave.
